File: apps/fogg/src/impl/version/get.go
package version

import (
	_ "embed"
	"fogg/src/pump"
	"fogg/src/pump/into"
	"fogg/src/type"
)

//go:embed get.json
var versionJSON string

var VersionGet types.Def

func init() {
	VersionGet = types.Def{
		Sid: into.StringToC("version"),
		Tag: into.StringToC("tag,any"),
		Fit: func(address types.Address, payload types.Payload, options types.Options) bool {
			if address == nil {
				return false
			}
			addr := into.CStr(address)
			return (addr == "/version" || addr == "version") && into.HasVerb(options, "GET")
		},
		Fun: func(address types.Address, payload types.Payload, options types.Options) int {
			if !VersionGet.Fit(address, payload, options) {
				return -1
			}
			target := into.GetInto(options)
			if target != nil {
				return pump.Pump(target, into.StringToC(versionJSON), into.StringToC("once"))
			}
			return 1
		},
	}
}

File: apps/fogr/src/impl/version/get.rs
#![allow(non_upper_case_globals)]

use core::ffi::CStr;
use crate::pump::into::{get_into, has_verb};
use crate::pump::Pump;
use crate::r#type::{Address, Def, Options, Payload};

pub static VersionGet: Def = Def {
    sid: c"version".as_ptr(),
    tag: c"tag,any".as_ptr(),
    fit: |address: Address, _payload: Payload, options: Options| -> bool {
        if address.is_null() { return false; }
        let addr = unsafe { CStr::from_ptr(address).to_bytes() };
        (addr == b"/version" || addr == b"version") && has_verb(options, b"GET")
    },
    fun: |address: Address, payload: Payload, options: Options| -> i32 {
        if !(VersionGet.fit)(address, payload, options) { return -1; }
        let json_payload = concat!(include_str!("get.json"), "\0").as_ptr() as Payload;
        if let Some(into_target) = get_into(options) {
            return Pump(into_target, json_payload, c"once".as_ptr());
        }
        1
    },
};
